### ISSUE TITLE
rolling with aggregation involving shift is very slow

### ISSUE THREAD

**Issue by 1112114641** [pola-rs/polars#15757]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import numpy as np
import polars as pl
from datetime import datetime

df = pl.DataFrame(
  {
    "date_working": pl.datetime_range(datetime(2020, 1, 1,0,0,0), datetime(2024, 3, 1,0,0,0), "1m",eager=True),
    "x": pl.int_range(0,2_190_241,step=1,eager=True),
  }
).with_columns(y=pl.col("x").sin()/1000+1).filter(pl.col("date_working").dt.weekday().le(4) & pl.col("date_working").dt.hour().ne(15))


def abc(y: pl.Expr) -> pl.Expr:
  y_pct = y.log() - y.log().shift()
  return y_pct.std(ddof=0).alias(f"{y.meta.output_name()}_abc")


df = df.with_columns(
  abc(pl.col("y")).rolling(index_column="date_working", period="2d")
)
df = df.with_columns(
  abc(pl.col("y")).rolling(index_column="date_working", period="2880i")
)

```

### Issue description

the version with ``period="2880i"`` runs in ~1.2s on my machine, the ``period="2d"`` has not finished after ~10min, and memory usage has increased to ~40GB for this tiny, tiny dataframe. After >15min, the kernel crashes as memory consumption went over what my machine can provision.

I have seen this behaviour for both 0.20.19 & 0.20.21

### Expected behaviour

Both versions should finish running in under 10s, given how "simple" the desired computation is.

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.21
Index type:           UInt32
Platform:             macOS-13.6.5-arm64-arm-64bit
Python:               3.10.6 (main, Aug  7 2023, 13:38:39) [Clang 14.0.3 (clang-1403.0.22.14.1)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          3.0.0
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           3.8.0
nest_asyncio:         1.5.8
numpy:                1.26.0
openpyxl:             3.1.2
pandas:               2.1.1
pyarrow:              13.0.0
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           2.0.22
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

**Comment by 1112114641** [comment]
@MarcoGorelli  - as promised in discord: a reproducible example

**Comment by MarcoGorelli** [comment]
thanks for the report!

First,
```python
df = df.with_columns(
  abc(pl.col("y")).rolling(index_column="date_working", period="2880i")
)
```
shouldn't work at all - `'i'` is meant for index counts, not for datetimes. https://github.com/pola-rs/polars/pull/15751 will disallow it.

---

As for why
```python
df = df.with_columns(
  abc(pl.col("y")).rolling(index_column="date_working", period="2d")
)
```
is slow, I think it's that the custom aggregation function you're doing is quite expensive

If we run an aggregation like:
```python
In [8]: %time df.with_columns((pl.col('y').log().mean()*(1-pl.col('y').sum())).rolling('date_working', period='2d'))
CPU times: user 51.3 ms, sys: 121 ms, total: 172 ms
Wall time: 89.4 ms
Out[8]:
shape: (1_200_600, 3)
┌─────────────────────┬─────────┬───────────┐
│ date_working        ┆ x       ┆ y         │
│ ---                 ┆ ---     ┆ ---       │
│ datetime[μs]        ┆ i64     ┆ f64       │
╞═════════════════════╪═════════╪═══════════╡
│ 2020-01-01 00:00:00 ┆ 0       ┆ 0.0       │
│ 2020-01-01 00:01:00 ┆ 1       ┆ -0.000421 │
│ 2020-01-01 00:02:00 ┆ 2       ┆ -0.001168 │
│ 2020-01-01 00:03:00 ┆ 3       ┆ -0.001419 │
│ 2020-01-01 00:04:00 ┆ 4       ┆ -0.000907 │
│ …                   ┆ …       ┆ …         │
│ 2024-02-29 23:55:00 ┆ 2190235 ┆ -0.001129 │
│ 2024-02-29 23:56:00 ┆ 2190236 ┆ -0.000596 │
│ 2024-02-29 23:57:00 ┆ 2190237 ┆ -0.001777 │
│ 2024-02-29 23:58:00 ┆ 2190238 ┆ -0.003586 │
│ 2024-02-29 23:59:00 ┆ 2190239 ┆ -0.00436  │
└─────────────────────┴─────────┴───────────┘
```
then it runs nearly instantly on your dataset

🤔 so what makes your example slow? It looks like it's the `y - y.shift` part, as the following takes 1.5s second on just 10_000 rows:
```python
In [25]: %time df.head(10_000).with_columns(y_abc=(pl.col('y')-pl.col('y').shift()).rolling(index_column="date_working", period="2d"))
CPU times: user 219 ms, sys: 1.29 s, total: 1.5 s
Wall time: 1.5 s
Out[25]:
shape: (10_000, 4)
┌─────────────────────┬───────┬──────────┬────────────────────────────────┐
│ date_working        ┆ x     ┆ y        ┆ y_abc                          │
│ ---                 ┆ ---   ┆ ---      ┆ ---                            │
│ datetime[μs]        ┆ i64   ┆ f64      ┆ list[f64]                      │
╞═════════════════════╪═══════╪══════════╪════════════════════════════════╡
│ 2020-01-01 00:00:00 ┆ 0     ┆ 1.0      ┆ [null]                         │
│ 2020-01-01 00:01:00 ┆ 1     ┆ 1.000841 ┆ [null, 0.000841]               │
│ 2020-01-01 00:02:00 ┆ 2     ┆ 1.000909 ┆ [null, 0.000841, 0.000068]     │
│ 2020-01-01 00:03:00 ┆ 3     ┆ 1.000141 ┆ [null, 0.000841, … -0.000768]  │
│ 2020-01-01 00:04:00 ┆ 4     ┆ 0.999243 ┆ [null, 0.000841, … -0.000898]  │
│ …                   ┆ …     ┆ …        ┆ …                              │
│ 2020-01-14 05:35:00 ┆ 19055 ┆ 0.999054 ┆ [null, -0.000162, … -0.000708] │
│ 2020-01-14 05:36:00 ┆ 19056 ┆ 0.999216 ┆ [null, -0.000162, … 0.000162]  │
│ 2020-01-14 05:37:00 ┆ 19057 ┆ 1.000099 ┆ [null, -0.000162, … 0.000883]  │
│ 2020-01-14 05:38:00 ┆ 19058 ┆ 1.000891 ┆ [null, -0.000162, … 0.000792]  │
│ 2020-01-14 05:39:00 ┆ 19059 ┆ 1.000864 ┆ [null, -0.000162, … -0.000027] │
└─────────────────────┴───────┴──────────┴────────────────────────────────┘
```

I don't know why `(pl.col('y')-pl.col('y').shift())` quite expensive to compute in this context

**Comment by MarcoGorelli** [comment]
## TL;DR

The following is slow:
```python
import polars as pl
from datetime import datetime

df = (
    pl.DataFrame(
        {
            "date_working": pl.datetime_range(
                datetime(2020, 1, 1, 0, 0, 0),
                datetime(2024, 3, 1, 0, 0, 0),
                "1m",
                eager=True,
            ),
            "x": pl.int_range(0, 2_190_241, step=1, eager=True),
        }
    )
    .with_columns(y=pl.col("x").sin() / 1000 + 1)
)

df.head(10_000).with_columns(
    y_abc=(pl.col("y") - pl.col("y").shift()).rolling(
        index_column="date_working", period="2d"
    )
)
```
Removing `.head(1000)` makes it hang indefinitely

@orlp @ritchie46 sorry for the ping, do you know why this aggregation might be slow here, compared with something like `(pl.col('y').log().mean()*(1-pl.col('y').sum())` which runs nearly instantly?

**Comment by orlp** [comment]
Almost surely because it's not shifting once and substracting once, but rather doing that for each window.

**Comment by MarcoGorelli** [comment]
Thanks both!

Indeed,
```python
def abc(y: pl.Expr, y_shifted) -> pl.Expr:
    y_pct = y.log() - y_shifted.log()
    return y_pct.std(ddof=0).alias(f"{y.meta.output_name()}_abc")


df.with_columns(y_shifted=pl.col("y").shift()).with_columns(
    abc(pl.col("y"), pl.col("y_shifted")).rolling(
        index_column="date_working", period="2d"
    )
)
```
run nearly instantly

(you could also pre-compute `.log()`, I just wanted to show the minimum needed to address the issue)

**Comment by orlp** [comment]
@ritchie46

> But our expression engine is conservative here as it doesn't know it can do the shifts for adjacent groups first.

That's not even correct by itself, you'd also need to make the first element in each group `null`, as that's what shifting does.

**Comment by ritchie46** [comment]
> That's not even correct by itself, you'd also need to make the first element in each group `null`, as that's what shifting does.

Good point. Then we don't have to do anything. ^^

I think we can close this. It is slow because it is expensive.
